app/controllers/api/v1/project_folders_controller.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class ProjectFoldersController < BaseController
+      before_action :load_team
+      before_action :load_project_folder, only: %i(show update)
+      before_action :check_parent_folder, only: %i(create update)
+
+      def index
+        project_folders = @team.project_folders
+                               .page(params.dig(:page, :number))
+                               .per(params.dig(:page, :size))
+
+        render jsonapi: project_folders, each_serializer: ProjectFolderSerializer, status: :ok
+      end
+
+      def show
+        render jsonapi: @project_folder, serializer: ProjectFolderSerializer, status: :ok
+      end
+
+      def create
+        folder = @team.project_folders.create!(folder_params)
+
+        render jsonapi: folder, serializer: ProjectFolderSerializer, status: :created
+      end
+
+      def update
+        @project_folder.attributes = update_folder_params
+
+        if @project_folder.changed? && @project_folder.save!
+          render jsonapi: @project_folder, serializer: ProjectFolderSerializer, status: :ok
+        else
+          render jsonapi: nil, status: :no_content
+        end
+      end
+
+      private
+
+      def load_project_folder
+        @project_folder = @team.project_folders.find(params.require(:id))
+      end
+
+      def folder_params
+        raise TypeError unless params.require(:data).require(:type) == 'project_folders'
+
+        params.from_jsonapi.require(:project_folder).permit(:name, :parent_folder_id)
```

Comment:
why do we need from_jsonapi here?

Reply:
As I mentioned in a meeting last week, `from_jsonapi` translates the raw json:api parameters into “classic” rails parameters, with a model top-level key and attributes in it’s object. The classic rails parameters are more conventional and easier to work with. As an added bonus, it handles relationships and parses them as `#{relationship_name}_id` or `#{relationship_name}_ids` for has_many relationships.